src/main/webapp/view/discussion/discussion-manager.js

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {COMMENT_TYPE_REPLY} from './discussion.js';
+
+/**
+ * Manages fetching and posting comments in a lecture's discussion.
+ */
+export default class DiscussionManager {
+  static #ENDPOINT = '/discussion';
+  static #PARAM_LECTURE = 'lecture';
+  static #PARAM_PARENT = 'parent';
+  static #PARAM_TIMESTAMP = 'timestamp';
+  static #PARAM_TYPE = 'type';
+  #lecture;
+
+  /**
+   * @param lecture The current lecture, which is used when making requests to
+   *     the discussion endpoint.
+   */
+  constructor(lecture) {
+    this.#lecture = lecture;
+  }
+
+  /**
+   * Organizes comments into threads with nested replies.
+   *
+   * <p>All comments are sent from the servlet without any structure, so the
+   * client needs to organize them before displaying.
+   */
+  structureComments(allComments) {
+    const commentIds = {};
+    for (const comment of allComments) {
+      comment.replies = [];
+      commentIds[comment.commentKey.id] = comment;
+    }
+
+    const rootComments = [];
+    for (const comment of allComments) {
+      if (comment.type === COMMENT_TYPE_REPLY) {
+        const parent = commentIds[comment.parentKey.value.id];
+        parent.replies.push(comment);
+      } else {
+        // Top level comments don't have parents.
+        rootComments.push(comment);
+      }
+    }
+    // Sort comments such that earliest timestamp is first.
+    rootComments.sort((a, b) => (a.timestampMs.value - b.timestampMs.value));
+    return rootComments;
+  }
+
+  /**
+   * Fetches and returns all comments in `this.#lecture` from the `ENDPOINT`.
```

Comment:
Methods should never reference internal variables. You can make this doc simpler:

Fetches all of the lectures comments from the `ENDPOINT` and returns them sorted by timestamp.

Or something like that.

Reply:
Changed, thanks!